contracts/lib/LibPrices.sol

```diff
@@ -0,0 +1,121 @@
+        uint256 cumulativePrice;
+        uint256 trades;
+    }
+
+    struct TWAP {
+        uint256 underlying;
+        uint256 derivative;
+    }
+
+    function fairPrice(uint256 oraclePrice, int256 timeValue)
+        public
+        pure
+        returns (uint256)
+    {
+        return uint256(LibMath.abs(oraclePrice.toInt256() - timeValue));
+    }
+
+    function timeValue(uint256 averageTracerPrice, uint256 averageOraclePrice)
+        public
+        pure
+        returns (int256)
+    {
+        return int256((averageTracerPrice - averageOraclePrice) / 90);
+    }
+
+    function averagePrice(PriceInstant memory price)
+        public
+        pure
+        returns (uint256)
+    {
+        return price.cumulativePrice / price.trades;
+    }
+
+    function averagePriceForPeriod(PriceInstant[] memory prices)
+        public
+        pure
+        returns (uint256)
+    {
+        uint256 n = prices.length <= 24 ? prices.length : 24;
+        uint256[] memory averagePrices = new uint256[](24);
+
+        for (uint256 i = 0; i < n; i++) {
+            PriceInstant memory currPrice = prices[i];
+            averagePrices[i] = averagePrice(currPrice);
+        }
+
+        return LibMath.mean(averagePrices);
+    }
+
+    function globalLeverage(
+        uint256 globalLeverage,
+        uint256 oldLeverage,
+        uint256 newLeverage
+    ) public pure returns (uint256) {
+        uint256 delta = oldLeverage - newLeverage;
+
+        if (oldLeverage < newLeverage) {
+            return globalLeverage + delta;
+        } else {
+            if (delta == globalLeverage) {
+                return 0;
+            } else {
+                return uint256(globalLeverage - delta);
+            }
+        }
+    }
+
+    function calculateTWAP(
+        uint256 hour,
+        PriceInstant[] memory tracerPrices,
+        PriceInstant[] memory oraclePrices
+    ) public pure returns (TWAP memory) {
+        uint256 instantDerivative = 0;
+        uint256 cumulativeDerivative = 0;
+        uint256 instantUnderlying = 0;
+        uint256 cumulativeUnderlying = 0;
+
+        for (uint256 i = 0; i < 8; i++) {
+            uint256 currTimeWeight = 8 - i;
+            uint256 j = 8 - i;
+
+            uint256 currDerivativePrice = averagePrice(tracerPrices[j]);
+            uint256 currUnderlyingPrice = averagePrice(oraclePrices[j]);
+
+            if (currDerivativePrice > 0) {
+                instantDerivative += currTimeWeight;
+                cumulativeDerivative += currTimeWeight * currDerivativePrice;
+            }
+
+            if (currUnderlyingPrice > 0) {
+                instantUnderlying += currTimeWeight;
+                cumulativeUnderlying += currTimeWeight * currUnderlyingPrice;
+            }
+
+            if (instantDerivative == 0) {
+                return TWAP(0, 0);
+            } else {
+                return
+                    TWAP(
+                        cumulativeUnderlying / instantUnderlying,
```

Comment:
These should be WAD divides since they are simply a scaled underlying average price, which is a WAD value. 

If we don't use WAD division, the returned value will not be a WAD and will throw out a lot.

Reply:
Consult 7d054b5.